agent/processwatcher.go

```diff
@@ -0,0 +1,434 @@
+	procLibs := make([]map[string]interface{}, len(libs))
+
+	for i, lib := range libs {
+		procLibs[i] = map[string]interface{}{
+			"outdated":      lib.Outdated,
+			"library_index": lib.libraryIndex,
+		}
+	}
+
+	return procLibs
+}
+
+func libToMap(lib libspector.Library) map[string]interface{} {
+	path := lib.Path()
+
+	modified, err := lib.Modified()
+	if err != nil {
+		log.Warningf("error retrieving modification date for lib %s, %v", path, err)
+		return nil
+	}
+
+	pkgName := ""
+	pkgVersion := ""
+	if pkg, err := lib.Package(); err != nil {
+		log.Warningf("error retrieving package name for lib %s, %v", path, err)
+	} else {
+		pkgName = pkg.Name()
+		pkgVersion = pkg.Version()
+	}
+
+	return map[string]interface{}{
+		"path":            path,
+		"modified":        modified,
+		"package_name":    pkgName,
+		"package_version": pkgVersion,
+	}
+}
+
+func (pm *processMap) findLibraryIndex(path string) int {
+	for i, library := range pm.libraries {
+		if library.Path() == path {
+			return i
+		}
+	}
+
+	return -1
+}
+
+func (pm *processMap) maybeAddLibrary(lib libspector.Library) int {
+	path := lib.Path()
+	index := pm.findLibraryIndex(path)
+
+	if index < 0 {
+		pm.libraries = append(pm.libraries, lib)
+		index = pm.findLibraryIndex(path)
+	}
+
+	return index
+}
+
+func (wt *processWatcher) acquireState() *processMap {
+	procs, err := wt.processes()
+	if err != nil {
+		panic(fmt.Errorf("Couldn't load processes: %s", err))
+	}
+
+	pm := processMap{
+		processes: make(systemProcesses, 0, len(procs)),
+		libraries: make(systemLibraries, 0), // ¯\_(ツ)_/¯
+	}
+
+	for _, proc := range procs {
+		started, err := proc.Started()
+		if err != nil {
+			log.Infof("PID %d is not running, skipping", proc.PID())
+			continue
+		}
+
+		command, err := proc.Command()
+		if err != nil {
+			log.Infof("Can't read command line for PID %d: %v", proc.PID(), err)
+			// fall through, we can live without this (?)
+		}
+
+		spectorLibs, err := proc.Libraries()
+		if err != nil {
+			if os.Getuid() != 0 && os.Geteuid() != 0 {
+				log.Infof("Cannot examine libs for PID %d, with UID:%d, EUID:%d",
+					proc.PID(), os.Getuid(), os.Geteuid())
+				continue
+			}
+
+			// otherwise barf
+			panic(fmt.Errorf("Couldn't load libs for process %v: %s", proc, err))
+		}
+
+		wp := watchedProcess{
+			ProcessStartedAt: started,
+			Pid:              proc.PID(),
+			Libraries:        make([]processLibrary, len(spectorLibs)),
+			Outdated:         false,
+			Command:          command,
+		}
+
+		for i, spectorLib := range spectorLibs {
+			libraryIndex := pm.maybeAddLibrary(spectorLib)
+			lib := processLibrary{
+				libraryIndex: libraryIndex,
+				Outdated:     spectorLib.Outdated(proc),
+			}
+			if lib.Outdated && !wp.Outdated {
+				wp.Outdated = true
+			}
+			wp.Libraries[i] = lib
+		}
+
+		pm.processes = append(pm.processes, wp)
+	}
+
+	// to make the crc check more meaningful
+	sort.Sort(pm.processes)
+	sort.Sort(pm.libraries)
```

Comment:
Good catch, a proper bug. I think I resolved it with a more thoughtful sort process.